opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py

```diff
@@ -733,7 +715,7 @@ def start_span(  # pylint: disable=too-many-locals
 
         if parent_context is None or not parent_context.is_valid:
             parent = parent_context = None
-            trace_id = generate_trace_id()
+            trace_id = self.source.ids_generator.generate_trace_id()
```

Comment:
Since this isn't part of the tracer provider API as far as I can tell from this PR so we can't guarantee that source will have an ids_generator attribute.

Reply:
That's a good point, but this is just being consistent with the assumption [later in this file](https://github.com/open-telemetry/opentelemetry-python/pull/1153/files/e6a970115c2512d0911efc23239e856c0ac71c10#diff-2f7b55162ad705df55ee48eb71914737R755) for the `sampler`, `resource`, and `_active_span_processor` variables.

Even so, I think we can guarantee the `source` will have an `ids_generator` attribute because source [is typed](https://github.com/open-telemetry/opentelemetry-python/pull/1153/files/e6a970115c2512d0911efc23239e856c0ac71c10#diff-2f7b55162ad705df55ee48eb71914737R677) to be of the same type as `TracerProvider` which will always have an `ids_generator` because of [the change](https://github.com/open-telemetry/opentelemetry-python/pull/1153/files/e6a970115c2512d0911efc23239e856c0ac71c10#diff-2f7b55162ad705df55ee48eb71914737R817) in this PR.

I think that the only danger that could come about would be if you were to implement your own `Tracer` class that expects an `ids_generator`, in which case you would also be in danger of assuming there were a `sampler` or `resource` as well because they're also not in the API.

Reply:
> Even so, I think we can guarantee the source will have an ids_generator attribute because source is typed to be of the same type as TracerProvider which will always have an ids_generator because of the change in this PR.

I don't think this guarantees anything really. It's only a type hint and makes mypy or other checkers complain if they are run. There is no runtime or build/package time guarantees.

> I think that the only danger that could come about would be if you were to implement your own Tracer class that expects an ids_generator, in which case you would also be in danger of assuming there were a sampler or resource as well because they're also not in the API.

Another instance would be if someone implements a custom TracerProvider to add some functionality but wanted to return instance of the stock Tracer. I think both cases should be supported without surprises. Looks like SDK Tracer has taken an implicit dependency on the SDK TracerProvider which shouldn't be the case. Users and vendors should be able to implement just one or the other without having to satisfy undocumented APIs.

I think we should either update the TracerProvider API to specify all these properties or go all in on DI and modify Tracer so that it requires all these dependencies to be explicitly specified by a tracer provider implementation. Meaning IDGenerator would be an initialization argument to the Tracer class. TracerProvider can pass it down explicitly. That way people can implement custom Tracer or custom TracerProvider without surprises. 

Since there are other instances of such implicit dependencies, may be we can do this in another PR but if all maintainers agree some solution now, perhaps we can start with IDGenerator now and follow up with the rest.

Reply:
> Looks like SDK Tracer has taken an implicit dependency on the SDK TracerProvider which shouldn't be the case.

Yeah we should probably remove source from both `Meter` and `Tracer` SDK implementations and then pass in some sort of "config" object upon creation. Tracer and TracerProvider shouldn't be coupled. But yeah, might be for another PR. @owais can you create an issue for this?

Reply:
I see what you're saying, yes it makes more sense to have `Tracer` and `TraceProvider` not be requirements of each other.

I would also prefer to have that as a separate PR, and minimize the new things this PR is trying to do. The PR is pretty straightforward the way it is right now, and that kind of refactor could be lost in the IDs Generator specific log messages.

I wouldn't mind helping with doing that separate PR either if we can decide on what we want to do!

Reply:
Sounds good to me. Tracking here: https://github.com/open-telemetry/opentelemetry-python/issues/1181